crates/symbolicator/src/services/shared_cache.rs

```diff
@@ -0,0 +1,346 @@
+                Err(anyhow!(
+                    "Bad GCS response for shared_cache: {}",
+                    LogError(&e)
+                ))
+            }
+            Err(_) => Err(Error::msg("Timeout from GCS for shared_cache")),
+        }
+    }
+}
+
+impl FilesystemSharedCacheConfig {
+    async fn fetch<W>(&self, key: &SharedCacheKey, writer: &mut W) -> Result<()>
+    where
+        W: AsyncWrite + Unpin,
+    {
+        let abspath = self.path.join(key.relative_path());
+        log::debug!("Fetching debug file from {}", abspath.display());
+        let mut file = File::open(abspath).await?;
+        io::copy(&mut file, writer).await?;
+        Ok(())
+    }
+
+    async fn store(&self, key: SharedCacheKey, contents: &[u8]) -> Result<()> {
+        let abspath = self.path.join(key.relative_path());
+        if let Some(parent_dir) = abspath.parent() {
+            fs::create_dir_all(parent_dir).await?;
+        }
+        fs::write(abspath, contents).await?;
+        Ok(())
+    }
+}
+
+/// Key for a shared cache item.
+#[derive(Debug, Clone)]
+pub struct SharedCacheKey {
+    /// The name of the cache.
+    pub name: CacheName,
+    /// The cache version.
+    pub version: u32,
+    /// The local cache key.
+    pub local_key: CacheKey,
+}
+
+impl SharedCacheKey {
+    /// The relative path of this cache key within a shared cache.
+    fn relative_path(&self) -> PathBuf {
+        // Note that this always pushes the version into the path, this is fine since we do
+        // not need any backwards compatibility with existing caches for the shared cache.
+        let mut path = PathBuf::new();
+        path.push(self.name.to_string());
+        path.push(self.version.to_string());
+        path.push(self.local_key.relative_path());
+        path
+    }
+
+    /// The [`SharedCacheKey::relative_path`] as a GCS bucket key.
+    fn gcs_bucket_key(&self) -> String {
+        // All our paths should be UTF-8, we don't construct non-UTF-8 paths.
+        match self.relative_path().to_str() {
+            Some(s) => s.to_owned(),
+            None => {
+                log::error!(
+                    "Non UTF-8 path in SharedCacheKey: {}",
+                    self.relative_path().display()
+                );
+                self.relative_path().to_string_lossy().into_owned()
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum SharedCacheBackend {
+    Gcs(GcsState),
+    Fs(FilesystemSharedCacheConfig),
+}
+
+/// A shared cache service.\
+///
+/// For simplicity in the rest of the application this service always exists, regardless of
+/// whether it is configured or not.  If it is not configured this becomes a no-op.
+#[derive(Debug)]
+pub struct SharedCacheService {
+    backend: Option<SharedCacheBackend>,
+}
+
+impl SharedCacheService {
+    pub fn new(config: Option<SharedCacheConfig>) -> Self {
+        // Yes, it is wasteful to create a GcsState when the config does not use GCS.
```

Comment:
Is this comment out of date? You are matching and then creating the state only when the config says its used.

Reply:
yes, outdated